File: php/luigi-template.php
<?php

namespace Luigi;

const VERSION = '0.4.0';

final class Error extends \Exception {};

final class Filters {
  private static $filters = null;

  public static function init() {
    if (self::$filters !== null)
      return;

    self::$filters = array(
      'h' => function($s) {
        return htmlspecialchars($v);
      },

      'u' => function($s) {
        return urlencode($v);
      },

      'json' => function($v) {
        return json_encode($v);
      },

      'hash' => function($v, $args) {
        $algo = (count($args) == 1) ? $args[0] : 'md5';
        return hash($algo, $v);
      },

      'base64' => function($v) {
        return base64_encode($v);
      },

      'nl2br' => function($v) {
        return nl2br($v);
      },

      'uc' => function($v) {
        return strtoupper($v);
      },

      'lc' => function($v) {
        return strtolower($v);
      },

      'trim' => function($v) {
        return trim($v);
      },

      'rtrim' => function($v) {
        return rtrim($v);
      },

      'ltrim' => function($v) {
        return ltrim($v);
      },

      's' => function($v) {
        return ($v == 1) ? '' : 's';
      },

      'strlen' => function($v) {
        return strlen($v);
      },

      'count' => function($v) {
        return count($v);
      },
    );
  }

  public static function get($key) {
    self::init();

    if (!isset(self::$filters[$key]))
      throw new Error("unknown filter: $key");

    return self::$filters[$key];
  }

  public static function add(array $filters) {
    self::init();
    self::$filters = array_merge(self::$filters, $filters);
  }
};

final class Parser {
  private static $RES = array(
    'action' => '/
      # match opening brace
      %\{
      
      # match optional whitespace
      \s*

      # match key
      (?<key>[^\s\|\}]+)

      # match filter(s)
      (?<filters>(\s*\|(\s*[^\s\|\}]+)+)*)

      # match optional whitespace
      \s*
    
      # match closing brace
      \}
      
      # or match up all non-% chars or a single % char
      | (?<text>[^%]* | %)
    /mx',

    'filter' => '/
      # match filter name
      (?<name>\S+)
    
      # match filter arguments (optional)
      (?<args>(\s*\S+)*)
    
      # optional trailing whitespace
      \s*
    /mx',

    'delim_filters' => '/\s*\|\s*/m',
    'delim_args' => '/\s+/m',
  );

  public static function parse_template($template) {
    # build list of matches
    $matches = array();
    $num_matches = preg_match_all(
      self::$RES['action'],
      $template,
      $matches,
      PREG_SET_ORDER
    );

    # check for error
    if ($num_matches === false)
      throw new Error("invalid template: $template");
  
    # walk over matches and build list of actions
    $r = array_map(function($m) {
      if ($m['key'] !== '') {
        #  key and filters
        return array(
          'type'    => 'action',
          'key'     => $m['key'],
          'filters' => self::parse_filters($m['filters']),
        );
      } else {
        # literal text
        return array(
          'type' => 'text',
          'text' => $m['text'],
        );
      }
    }, $matches);

    # return result
    return $r;
  }

  public static function parse_filters($filters) {
    # split into individual filters
    $r = array();
    foreach (preg_split(self::$RES['delim_filters'], $filters) as $f) {
      # trim whitespace
      $f = trim($f);

      # skip empty filters
      if (!$f)
        continue;

      # match filter
      $md = array();
      if (!preg_match(self::$RES['filter'], $f, $md))
        throw new Error("invalid filter: $f");

      # add filter to results
      $r[] = array(
        # filter name
        'name' => $md['name'],

        # filter arguments
        'args' => (count($md) > 2) ? preg_split(
          self::$RES['delim_args'],
          trim($md['args'])
        ) : array(),
      );
    }

    # return results
    return $r;
  }
};

final class Template {
  private $template, $actions, $o;

  public function __construct($template, array $o = array()) {
    $this->template = $template;
    $this->o = $o;

    # parse template into list of actions
    $this->actions = Parser::parse_template($template);
  }

  public function run(array $args = array()) {
    # php sucks
    $me = $this;

    return join('', array_map(function($row) use ($me, $args) {
      if ($row['type'] == 'text') {
        # literal text
        return $row['text'];
      } else if ($row['type'] == 'action') {
        # template key (possibly with filters)

        # check value
        if (!isset($args[$row['key']]))
          throw new Error("unknown key: {$row['key']}");

        # pass value through filters and return result
        return array_reduce($row['filters'], function($r, $f) use ($me, $args) {
          # get filter
          $fn = Filters::get($f['name']);
        
          # call filter and return result
          return call_user_func($fn, $r, $f['args'], $args, $me);
        }, $args[$row['key']]);
      } else {
        # should never be reached
        throw new Error("unknown action type: {$row['type']}");
      }
    }, $this->actions));
  }

  public static function run_once($str, array $args = array()) {
    $t = new Template($str);
    return $t->run($args);
  }
};

final class Cache {
  private $templates, $o, $lut = array();

  public function __construct(array $templates, array $o = array()) {
    $this->templates = $templates;
    $this->o = $o;
  }

  public function get($key) {
    if (!isset($this->lut[$key])) {
      if (!isset($this->templates[$key]))
        throw new Error("unknown template: $key");

      # lazy-load template
      $this->lut[$key] = new Template($this->templates[$key], $this->o);
    }

    # return result
    return $this->lut[$key];
  }

  public function run($key, array $args = array()) {
    return $this->get($key)->run($args);
  }
};

File: ruby/luigi-template.rb
# require 'pp'

module Luigi
  #
  # library version
  #
  VERSION = '0.4.0'

  #
  # built-in filters
  #
  FILTERS = {
    # upper-case string
    uc: proc { |v, args, row, t|
      (v || '').to_s.upcase
    },

    # lower-case string
    lc: proc { |v, args, row, t|
      (v || '').to_s.downcase
    },

    # html-escape string
    h: proc { |v, args, row, t|
      (v || '').to_s.gsub(/&/, '&amp;').gsub(/</, '&lt;').gsub(/>/, '&gt;').gsub(/'/, '&apos').gsub(/"/, '&quot;')
    },

    # uri-escape string
    u: proc( { |v, args, row, t|
      require 'uri'
      URI.escape((v || '').to_s)
    },

    # json-encode value
    json: proc { |v, args, row, t|
      require 'json'
      v.to_json
    },

    # trim leading and trailing whitespace from string
    trim: proc { |v, args, row, t|
      (v || '').to_s.strip
    },

    # base64-encode string
    base64: proc { |v, args, row, t|
      [(v || '').to_s].pack('m')
    },

    # hash string
    hash: proc { |v, args, row, t|
      require 'openssl'
      OpenSSL::Digest.new(args[0] || 'md5').hexdigest((v || '').to_s)
    },
  }

  #
  # Template parser.
  #
  module Parser
    RES = {
      action: %r{
        # match opening brace
        %\{

        # match optional whitespace
        \s*

        # match key
        (?<key>[^\s\|\}]+)

        # match filter(s)
        (?<filters>(\s*\|(\s*[^\s\|\}]+)+)*)

        # match optional whitespace
        \s*

        # match closing brace
        \}

        # or match up all non-% chars or a single % char
        | (?<text>[^%]* | %)
      }mx,

      filter: %r{
        # match filter name
        (?<name>\S+)

        # match filter arguments (optional)
        (?<args>(\s*\S+)*)

        # optional trailing whitespace
        \s*
      }mx,

      delim_filters: %r{
        \s*\|\s*
      }mx,

      delim_args: %r{
        \s+
      },
    }

    #
    # Parse a (possibly empty) string into an array of actions.
    #
    def self.parse_template(str)
      str.scan(RES[:action]).map { |m|
        if m[0] && m[0].length > 0
          r = {
            type: :action,
            key: m[0].intern,
            filters: parse_filters(m[1]),
          }
        else
          # literal text
          r = { type: :text, text: m[2] }
        end

        # pp r

        # return result
        r
      }
    end

    #
    # Parse a (possibly empty) string into an array of filters.
    #
    def self.parse_filters(str)
      # strip leading and trailing whitespace
      str = (str || '').strip

      if str.length > 0
        str.strip.split(RES[:delim_filters]).inject([]) do |r, f|
          # strip whitespace
          f = f.strip

          if f.length > 0
            md = f.match(RES[:filter])
            raise "invalid filter: #{f}" unless md
            # pp md

            # get args
            args = md[:args].strip

            # add to result
            r << {
              name: md[:name].intern,
              args: args.length > 0 ? args.split(RES[:delim_args]) : [],
            }
          end

          # return result
          r
        end
      else
        # return empty filter set
        []
      end
    end
  end

  #
  # Template class.
  #
  class Template
    #
    # Create a new Template from the given string.
    #
    def initialize(str, filters = FILTERS)
      @str, @filters = str, filters
      @actions = Parser.parse_template(str)
    end

    #
    # Run template with given arguments
    #
    def run(args)
      @actions.map { |a|
        # pp a

        case a[:type]
        when :action
          # check key and get value
          val = if args.key?(a[:key])
            args[a[:key]]
          elsif args.key?(a[:key].to_s)
            args[a[:key].to_s]
          else
            # invalid key
            raise "unknown argument: #{a[:key]}"
          end

          # filter value
          a[:filters].inject(val) do |r, f|
            # check filter name
            raise "unknown filter: #{f[:name]}" unless @filters.key?(f[:name])

            # call filter, return result
            @filters[f[:name]].call(r, f[:args], args, self)
          end
        when :text
          # literal text
          a[:text]
        else
          # never reached
          raise "unknown action type: #{a[:type]}"
        end
      }.join
    end
  end

  #
  # Simple template cache.
  #
  class Cache
    #
    # Create a new template cache with the given templates
    #
    def initialize(strings, filters = FILTERS)
      @templates = Hash.new do |h, k|
        # always deal with symbols
        k = k.intern

        # make sure template exists
        raise "unknown template: #{k}" unless strings.key?(k)

        # create template
        h[k] = Template.new(strings[k], filters)
      end
    end

    #
    # Run specified template with given arguments.
    #
    def run(key, args)
      # run template with args and return result
      @templates[key].run(args)
    end
  end

  #
  # test module
  #
  module Test
    # test template
    TEMPLATE = '
      basic test: hello %{name}
      test filters: hello %{name | uc | base64 | hash sha1}
      test custom: hello %{name|custom}
      test custom_with_arg: %{name|custom_with_arg asdf}
    '

    CUSTOM_FILTERS = {
      custom: proc {
        'custom'
      },

      custom_with_arg: proc { |v, args|
        args.first || 'custom'
      },
    }

    # test template cache
    CACHE = {
      test: TEMPLATE
    }

    # test arguments
    ARGS = {
      name: 'paul',
    }

    def self.run
      # add custom filters
      Luigi::FILTERS.update(CUSTOM_FILTERS)

      # test individual template
      puts Template.new(TEMPLATE).run(ARGS)

      # test template cache
      puts Cache.new(CACHE).run(:test, ARGS)
    end
  end
end

Luigi::Test.run if __FILE__ == $0